content/chapters/compute/lecture/slides.mdpp

```diff
@@ -0,0 +1,1088 @@
+
+``` [1 - 2 | 3 - 4 | 5 - 6 | 7 - 8]
+student@os:~/.../compute/lecture/demo/race-condition$ ./race_condition
+var = 1232611; time = 14 ms
+student@os:~/.../compute/lecture/demo/race-condition$ ./race_condition_atomic
+var = 2000000; time = 31 ms
+student@os:~/.../compute/lecture/demo/race-condition$ ./race_condition_spinlock
+var = 2000000; time = 73 ms
+student@os:~/.../compute/lecture/demo/race-condition$ ./race_condition_mutex
+var = 2000000; time = 98 ms
+```
+
+* Why is the spinlock faster?
+    * **Short** critical section, so the lock is acquired fast
+    * The context switches performed by the mutex are costlier
+
+----
+
+## Critical Section - Granularity
+
+* What if we place all the code in `increase_var` in a critical section?
+* Short critical sections mean more frequent context switches
+* Large critical sections mean less parallelism (code runs almost sequentially)
+
+```
+student@os:~/.../compute/lecture/demo/granularity$ ./coarse_granularity
+var = 2000000; time = 36 ms
+student@os:~/.../compute/lecture/demo/granularity$ ./fine_granularity
+var = 2000000; time = 1268 ms
+```
+
+----
+
+## Who Should Release the Lock?
+
+![Locks vs Notifications](./media/lock-undefined-behaviour.svg)
+
+---
+
+## Locks vs Notifications
+
+![Lock vs Notification](./media/lock-vs-notify.svg)
+
+----
+
+## Locks vs Notifications
+
+* Locks:
+    * isolate **critical sections**
+    * thread that calls `unlock()` must hold the lock
+
+* Notifications (conditions and semaphores):
+    * one thread waits for an **event**
+    * **another thread** notifies waiting thread on event completion
+
+----
+
+### Semaphore
+
+* Generalized mutex (mutex = binary semaphore)
+* Lock can take values greater than 1
+* `acquire_lock() ~ lock--`:
+    * Threads enter critical section if `lock > 0`
+    * Threads block if `lock == 0`
+* `release_lock() ~ lock++`
+
+----
+
+### Spinlock vs Mutex vs Semaphore
+
+| Spinlock                                 | Mutex                                    | Semaphore                                   |
+| :--------------------------------------- | :--------------------------------------- | :------------------------------------------ |
+| Busy waiting                             | Yields CPU                               | Yields CPU                                  |
+| Good for short critical sections         | Good for complex critical sections       | Good for complex critical sections          |
+| Only 1 thread can enter critical section | Only 1 thread can enter critical section | Multiple threads can enter critical section |
+
+---
+
+### Deadlock
+
+* Circular dependency
+* Threads cyclically wait for each other
+
+![Deadlock](media/deadlock.svg)
+
+```
+student@os:~/.../compute/lecture/demo/deadlock$ python3 deadlock.py
+Thread 1: locked resource 1. Waiting for resource 2...
+Thread 2: locked resource 2. Waiting for resource 1...
+
+```
+
+* Now the threads are deadlocked
```

Comment:
The line is not properly displayed.

Reply:
I've removed it as it just states the obvious.